---
layout: post
title: 스프링 DB 1편 - 데이터 접근 핵심 원리 - 5. 자바 예외 이해
tags: [Spring, SpringDB1, UF]
permalink: /docs/Spring/SpringDB1_4
date: 2022-10-13 21:00:00
---
# 자바 예외 이해
## 예외 계층

**예외 계층 그림**
![자바 예외 이해 - 01  예외 계층](https://user-images.githubusercontent.com/52024566/195851243-59d311da-a993-4a1c-a17f-f6a8adf8aff8.png)

- `Object`: 예외도 객체. 모든 객체의 최상위 부모는 `Object`이므로 예외의 최상위 부모도 `Object`
- `Throwable`: 최상위 예외. 하위에 `Exception`과 `Error`가 있음
- `Error`: 메모리 부족이나 심각한 시스템 오류와 같이 애플리케이션에서 복구 불가능한 시스템 예외. 애플리케이션 개발자는 이 예외를 잡으려고 해서는 안됨
  - 상위 예외를 `catch`로 잡으면 그 하위 예외까지 함께 잡음. 따라서 애플리케이션 로직에서는 `Throwable` 예외도 잡으면 안되는데, 앞서 이야기한 `Error` 예외도 함께 잡을 수 있기 때문. 애플리케이션 로직은 이런 이유로 `Exception`부터 필요한 예외로 생각하고 잡으면 됨
  - 참고로 `Error`도 언체크 예외
- `Exception`: 체크 예외
  - 애플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외
  - `Exception`과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외. 단 `RuntimeException`은 예외
- `RuntimeException`: 언체크 예외, 런타임 예외
  - 컴파일러가 체크 하지 않는 언체크 예외
  - `RuntimeException`과 그 자식 예외는 모두 언체크 예외
  - `RuntimeException`의 이름을 따라서 `RuntimeException`과 그 하위 언체크 예외를 런타임 예외라고 함. 

## 예외 기본 규칙
예외는 폭탄 돌리기와 같음. 잡아서 처리하거나, 처리할 수 없으면 밖으로 던져야함

**예외 처리**
![자바 예외 이해 - 02  예외 처리](https://user-images.githubusercontent.com/52024566/195851248-303e30e8-fc71-43f9-9264-ffad354d802b.png)

- 5번에서 예외를 처리하면 이후에는 애플리케이션 로직이 정상 흐름으로 동작

**예외 던짐**
![자바 예외 이해 - 03  예외 던짐](https://user-images.githubusercontent.com/52024566/195851253-8668f9d1-8750-4965-818a-a8da4750120a.png)

예외를 처리하지 못하면 호출한 곳으로 예외를 계속 던짐

**예외의  2가지 기본 규칙**
1. 예외는 잡아서 처리하거나 던져야 함
2. 예외를 잡거나 던질 때 지정한 예외뿐만 아니라 그 예외의 자식들도 함께 처리됨
  - 예를 들어서 `Exception`을 `catch`로 잡으면 그 하위 예외들도 모두 잡을 수 있음
  - 예를 들어서 `Exception`을 `throws`로 던지면 그 하위 예외들도 모두 던질 수 있음

**참고: 예외를 처리하지 못하고 계속 던지면?**
- 자바 `main()` 쓰레드의 경우 예외 로그를 출력하면서 시스템이 종료
- 웹 애플리케이션의 경우 여러 사용자의 요청을 처리하기 때문에 하나의 예외 때문에 시스템이 종료되면 안됨. WAS가 해당 예외를 받아서 처리하는데, 주로 사용자에게 개발자가 지정한 오류 페이지를 보여줌

## 체크 예외 기본 이해

- `Exception`과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외. 단 `RuntimeException`은 체크 예외가 아님
- 체크 예외는 잡아서 처리하거나, 또는 밖으로 던지도록 선언해야 함. 그렇지 않으면 컴파일 오류가 발생

**체크 예외 전체 코드**
```java
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.*;

@Slf4j
public class CheckedTest {
  
    @Test
    void checked_catch() {
        Service service = new Service();
        service.callCatch();
    }
  
    @Test
    void checked_throw() {
        Service service = new Service();
        assertThatThrownBy(() -> service.callThrow())
          .isInstanceOf(MyCheckedException.class);
    }
  
    /**
    * Exception을 상속받은 예외는 체크 예외가 된다.
    */
    static class MyCheckedException extends Exception {
        public MyCheckedException(String message) {
            super(message);
        }
    }
  
    /**
    * Checked 예외는
    * 예외를 잡아서 처리하거나, 던지거나 둘중 하나를 필수로 선택해야 한다.
    */
    static class Service {
        Repository repository = new Repository();
      
        /**
        * 예외를 잡아서 처리하는 코드
        */
        public void callCatch() {
            try {
                repository.call();
            } catch (MyCheckedException e) {
                //예외 처리 로직
                log.info("예외 처리, message={}", e.getMessage(), e);
            }
        }
      
        /**
        * 체크 예외를 밖으로 던지는 코드
        * 체크 예외는 예외를 잡지 않고 밖으로 던지려면 throws 예외를 메서드에 필수로
        선언해야한다.
        */
        public void callThrow() throws MyCheckedException {
            repository.call();
        }
    }
  
    static class Repository {
        public void call() throws MyCheckedException {
            throw new MyCheckedException("ex");
        }
    }
}
```

**Exception을 상속받은 예외는 체크 예외**
```java
static class MyCheckedException extends Exception {
public MyCheckedException(String message) {
super(message);
}
}
```

- `MyCheckedException`는 `Exception`을 상속받음. `Exception`을 상속받으면 체크 예외가 됨
- 참고로 `RuntimeException`을 상속받으면 언체크 예외가 됨. 이런 규칙은 자바 언어에서 문법으로 정함
- 예외가 제공하는 여러가지 기본 기능이 있는데, 그 중에 오류 메시지를 보관하는 기능도 있음. 예제에서 보는 것처럼 생성자를 통해서 해당 기능을 그대로 사용하면 편리

```java
@Test
void checked_catch() {
    Service service = new Service();
    service.callCatch();
}
```
- `service.callCatch()`에서 예외를 처리했기 때문에 테스트 메서드까지 예외가 올라오지 않음

실행 순서 분석
1. `test` → `service.callCatch()` → `repository.call()` [예외 발생, 던짐]
2. `test` ← `service.callCatch()` [예외 처리] ← `repository.call()`
3. `test` [정상 흐름] ← `service.callCatch()` ← `repository.call()`

`Repository.call()`에서 `MyUncheckedException` 예외가 발생하고, 그 예외를 `Service.callCatch()`에서 잡는 것을 확인할 수 있음
```java
log.info("예외 처리, message={}", e.getMessage(), e);
```

**실행 결과**
```
[Test worker] INFO hello.jdbc.exception.basic.UncheckedTest - 예외 처리,
message=ex
hello.jdbc.exception.basic.UncheckedTest$MyUncheckedException: ex
at
hello.jdbc.exception.basic.UncheckedTest$Repository.call(UncheckedTest.java:66)
at
hello.jdbc.exception.basic.UncheckedTest$Service.callCatch(UncheckedTest.java:4
8)
at
hello.jdbc.exception.basic.UncheckedTest.unchecked_catch(UncheckedTest.java:14)
```

- 실행 결과 로그를 보면 첫줄은 우리가 남긴 로그가 그대로 남는 것을 확인할 수 있음
- 그런데 두 번째 줄 부터 예외에 대한 스택 트레이스가 추가로 출력
- 이 부분은 로그를 남길 때 로그의 마지막 인수에 예외 객체를 전달해주면 로그가 해당 예외의 스택 트레이스를 추가로 출력
  - `log.info("예외 처리, message={}", e.getMessage(), e);` ← 여기서 마지막에 있는 `e`

**체크 예외를 잡아서 처리하는 코드**
```java
try {
    repository.call();
} catch (MyCheckedException e) {
    //예외 처리 로직
}
```
- 체크 예외를 잡아서 처리하려면 `catch(..)`를 사용해서 예외를 잡음
- 여기서는 `MyCheckedException` 예외를 잡아서 처리

**catch는 해당 타입과 그 하위 타입을 모두 잡을 수 있음**
```java
public void callCatch() {
    try {
        repository.call();
    } catch (Exception e) {
        //예외 처리 로직
    }
}
```

- `catch`에 `MyCheckedException`의 상위 타입인 `Exception`을 적어주어도 `MyCheckedException`을 잡을 수 있음
- `catch`에 예외를 지정하면 해당 예외와 그 하위 타입 예외를 모두 잡아줌
- 물론 정확하게 `MyCheckedException`만 잡고 싶다면 `catch`에 `MyCheckedException`을 적어주어야 함

**예외를 처리하지 않고 밖으로 던지는 코드**
```java
@Test
void checked_throw() {
    Service service = new Service();
    assertThatThrownBy(() -> service.callThrow())
      .isInstanceOf(MyCheckedException.class);
}
```

- `service.callThrow()`에서 예외를 처리하지 않고, 밖으로 던졌기 때문에 예외가 테스트 메서드까지 올라옴
- 테스트에서는 기대한 것처럼 `MyCheckedException` 예외가 던져지면 성공으로 처리

실행 순서 분석
1. `test` → `service.callThrow()` → `repository.call()` [예외 발생, 던짐]
2. `test` ← `service.callThrow()` [예외 던짐] ← `repository.call()`
3. `test` [예외 도착] ← `service.callThrow()` ← `repository.call()`

**체크 예외를 밖으로 던지는 코드**
```java
public void callThrow() throws MyCheckedException {
    repository.call();
}
```

- 체크 예외를 처리할 수 없을 때는 `method() throws 예외`을 사용해서 밖으로 던질 예외를 필수로 지정해주어야 함. 여기서는 `MyCheckedException`을 밖으로 던지도록 지정

**체크 예외를 밖으로 던지지 않으면 컴파일 오류 발생**
```java
public void callThrow() {
    repository.call();
}
```

- `throws`를 지정하지 않으면 컴파일 오류가 발생
  - `Unhandled exception: hello.jdbc.exception.basic.CheckedTest.MyCheckedException`
  - 체크 예외의 경우 예외를 잡아서 처리하거나 또는 `throws`를 지정해서 예외를 밖으로 던진다는 선언을 필수로 해주어야 함

**체크 예외를 밖으로 던지는 경우에도 해당 타입과 그 하위 타입을 모두 던질 수 있음**
```java
public void callThrow() throws Exception {
    repository.call();
}
```

- `throws`에 `MyCheckedException`의 상위 타입인 `Exception`을 적어주어도 `MyCheckedException`을 던질 수 있음
- `throws`에 지정한 타입과 그 하위 타입 예외를 밖으로 던짐
- 물론 정확하게 `MyCheckedException`만 밖으로 던지고 싶다면 `throws`에 `MyCheckedException`을 적어주어야 함

### 체크 예외의 장단점
체크 예외는 예외를 잡아서 처리할 수 없을 때, 예외를 밖으로 던지는 throws 예외 를 필수로 선언해야 함. 그렇지 않으면 컴파일 오류가 발생. 이것 때문에 장점과 단점이 동시에 존재
- 장점: 개발자가 실수로 예외를 누락하지 않도록 컴파일러를 통해 문제를 잡아주는 훌륭한 안전 장치
- 단점: 하지만 실제로는 개발자가 모든 체크 예외를 반드시 잡거나 던지도록 처리해야 하기 때문에, 너무 번거로운 일이 됨. 크게 신경쓰고 싶지 않은 예외까지 모두 챙겨야 함. 추가로 의존관계에 따른 단점도 있음

## 언체크 예외 기본 이해

- `RuntimeException`과 그 하위 예외는 언체크 예외로 분류됨
- 언체크 예외는 말 그대로 컴파일러가 예외를 체크하지 않는다는 뜻
- 언체크 예외는 체크 예외와 기본적으로 동일하나 차이가 있다면 예외를 던지는 `throws`를 선언하지 않고 생략할 수 있음. 이 경우 자동으로 예외를 던짐

### 체크 예외 VS 언체크 예외
- 체크 예외: 예외를 잡아서 처리하지 않으면 항상 `throws`에 던지는 예외를 선언해야 함
- 언체크 예외: 예외를 잡아서 처리하지 않아도 `throws`를 생략할 수 있음

**언체크 예외 전체 코드**
```java
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

@Slf4j
public class UncheckedTest {
  
    @Test
    void unchecked_catch() {
        Service service = new Service();
        service.callCatch();
    }
  
    @Test
    void unchecked_throw() {
        Service service = new Service();
        assertThatThrownBy(() -> service.callThrow())
          .isInstanceOf(MyUncheckedException.class);
    }
  
    /**
    * RuntimeException을 상속받은 예외는 언체크 예외가 된다.
    */
    static class MyUncheckedException extends RuntimeException {
        public MyUncheckedException(String message) {
            super(message);
        }
    }
  
    /**
    * UnChecked 예외는
    * 예외를 잡거나, 던지지 않아도 된다.
    * 예외를 잡지 않으면 자동으로 밖으로 던진다.
    */
    static class Service {
      
        Repository repository = new Repository();
      
        /**
        * 필요한 경우 예외를 잡아서 처리하면 된다.
        */
        public void callCatch() {
            try {
                repository.call();
            } catch (MyUncheckedException e) {
                //예외 처리 로직
                log.info("예외 처리, message={}", e.getMessage(), e);
            }
        }
      
        /**
        * 예외를 잡지 않아도 된다. 자연스럽게 상위로 넘어간다.
        * 체크 예외와 다르게 throws 예외 선언을 하지 않아도 된다.
        */
        public void callThrow() {
            repository.call();
        }
    }
  
    static class Repository {
        public void call() {
            throw new MyUncheckedException("ex");
        }
    }
}
```

**언체크 예외를 잡아서 처리하는 코드**
```java
try {
    repository.call();
} catch (MyUncheckedException e) {
    //예외 처리 로직
    log.info("error", e);
}
```

- 언체크 예외도 필요한 경우 이렇게 잡아서 처리할 수 있음

**언체크 예외를 밖으로 던지는 코드 - 생략**
```java
public void callThrow() {
    repository.call();
}
```

- 언체크 예외는 체크 예외와 다르게 `throws 예외`를 선언하지 않아도 됨
- 말 그대로 컴파일러가 이런 부분을 체크하지 않기 때문에 언체크 예외

**언체크 예외를 밖으로 던지는 코드 - 선언**
```java
public void callThrow() throws MyUncheckedException {
    repository.call();
}
```

- 참고로 언체크 예외도 `throws 예외`를 선언해도 됨. 물론 생략할 수 있음
- 언체크 예외는 주로 생략하지만, 중요한 예외의 경우 이렇게 선언해두면 해당 코드를 호출하는 개발자가 이런 예외가 발생한다는 점을 IDE를 통해 좀 더 편리하게 인지할 수 있음 (컴파일 시점에 막을 수 있는 것은 아니고, IDE를 통해서 인지할 수 있는 정도)

### 언체크 예외의 장단점
언체크 예외는 예외를 잡아서 처리할 수 없을 때, 예외를 밖으로 던지는 `throws 예외`를 생략할 수 있음. 이것 때문에 장점과 단점이 동시에 존재
- 장점: 신경쓰고 싶지 않은 언체크 예외를 무시할 수 있음. 체크 예외의 경우 처리할 수 없는 예외를 밖으로 던지려면 항상 `throws 예외`를 선언해야 하지만, 언체크 예외는 이 부분을 생략할 수 있음. 신경쓰고 싶지 않은 예외의 의존관계를 참조하지 않아도 되는 장점이 있음
- 단점: 언체크 예외는 개발자가 실수로 예외를 누락할 수 있음. 반면에 체크 예외는 컴파일러를 통해 예외 누락을 잡아줌

### 정리
체크 예외와 언체크 예외의 차이는 예외를 처리할 수 없을 때 예외를 밖으로 던지는 부분에 있음. 이 부분을 필수로 선언해야 하는가 생략할 수 있는가의 차이

## 체크 예외 활용
## 언체크 예외 활용
## 예외 포함과 스택 트레이스